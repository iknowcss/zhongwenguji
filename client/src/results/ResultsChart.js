import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Chart } from 'chart.js';
import style from './ResultsChart.module.scss';
import colors from '../style/colors.scss';

const lightTransparent = 'rgba(' + [].slice.call(colors.light.match(/^#?(..)(..)(..)$/), 1, 4).map(n => parseInt(n, 16)).join(',') + ',.5)';

function datify(curvePoints, samplePoints) {
  let stopIndex = 0;
  curvePoints.find(([,y], i) => (stopIndex = i) && y === 0);
  stopIndex = stopIndex + 2;

  return {
    datasets: [
      {
        data: Array.from(samplePoints, ([x, y]) => ({x, y})),
        backgroundColor: 'transparent',
        borderColor: colors.accent,
        showLine: false,
        borderWidth: 2,
        pointRadius: 2
      },
      {
        showLine: true,
        data: Array.from(curvePoints.slice(0, stopIndex), ([x, y]) => ({x, y})),
        backgroundColor: lightTransparent,
        borderColor: colors.primary,
        borderWidth: 2,
        pointRadius: 0
      }
    ]
  };
}

export const renderXTick = x => Math.round(x / 100) * 100;

export const renderYTick = y => y === 100 ? '100%' : y < 100 ? `${y}` : '';

export default class ResultsChart extends Component {
  static propTypes = {
    curvePoints: PropTypes.array,
    samplePoints: PropTypes.array,
  };

  static defaultProps = {
    curvePoints: [],
    samplePoints: []
  };

  constructor() {
    super();
    this.canvasRef = React.createRef();
  }

  destroyChart() {
    if (this.chart) {
      this.chart.destroy();
    }
    this.chart = null;
    this.canvasContext = null;
  }

  initChart() {
    const { curvePoints, samplePoints } = this.props;

    this.canvasContext = this.canvasRef.current.getContext('2d');
    this.chart = new Chart(this.canvasContext, {
      responsive: true,
      type: 'scatter',
      data: datify(curvePoints, samplePoints),
      options: {
        legend: { display: false },
        scales: {
          xAxes: [{
            ticks: {
              callback: renderXTick,
              maxTicksLimit: 5,
              startAtZero: true
            }
          }],
          yAxes: [{
            ticks: {
              callback: renderYTick,
              beginAtZero: true,
              max: 120
            }
          }]
        }
      }
    });
  }

  componentDidMount() {
    this.initChart();
  }

  componentWillUnmount() {
    this.destroyChart();
  }

  render() {
    return (
      <div className={style.container}>
        <canvas ref={this.canvasRef} />
      </div>
    )
  }
}
