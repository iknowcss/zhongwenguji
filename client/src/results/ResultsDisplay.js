import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import I18n from '../i18n/I18n';
import style from './ResultsDisplay.module.scss';

const ResultsDisplay = ({ estimate, uncertainty, className }) => (
  <div className={cx(style.container, className)}>
    <I18n
      component="div"
      stringId="results.youKnow"
      className={style.youKnow}
    />
    <div className={style.estimate}>
      <span className={style.mean}>{estimate}</span>
      <span className={style.plusMinus}>Â±</span>
      <span className={style.std}>{uncertainty}</span>
    </div>
    <I18n
      component="div"
      stringId="results.measure"
      className={style.measure}
    />
  </div>
);

ResultsDisplay.propTypes = {
  estimate: PropTypes.number,
  uncertainty: PropTypes.number
};

ResultsDisplay.defaultProps = {
  estimate: 0,
  resultData: 0
};

export default ResultsDisplay;