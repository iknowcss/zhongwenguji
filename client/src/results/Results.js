import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import cx from 'classnames';
import getConfig from '../getConfig';
import mapSelectors from '../util/mapSelectors';
import noop from '../util/noop';
import ResultsDisplay from './ResultsDisplay';
import ResultsChart from './ResultsChart';
import MissedCards from './MissedCards';
import { resultData, scoreStatistics, missedCards } from '../characterTest/characterTestReducer';
import { resetTest } from '../characterTest/characterTestActions';
import Button from '../component/Button';
import I18n from '../i18n/I18n';
import style from './Results.module.scss';

const feedbackPrefillUrl = getConfig().feedbackPrefillUrl;

class Results extends Component {
  static propTypes = {
    scoreStatistics: PropTypes.object,
    resultData: PropTypes.object,
    resetTest: PropTypes.func
  };

  static defaultProps = {
    scoreStatistics: {},
    resultData: {},
    resetTest: noop
  };

  handleStartAgainClick = (event) => {
    event.preventDefault();
    this.props.resetTest();
  };

  render() {
    const resultData = this.props.resultData || {};

    return (
      <div className={style.container}>
        <ResultsDisplay
          className={style.resultsDisplay}
          estimate={resultData.knownEstimate}
          uncertainty={resultData.knownEstimateUncertainty}
        />
        <ResultsChart
          curvePoints={resultData.curvePoints}
          samplePoints={resultData.samplePoints}
        />
        <div className={style.actionContainer}>
          <I18n component={Button} className={style.action} stringId="results.reviewMissedCharacters" />
          <I18n component={Button} className={style.action} stringId="results.startAgain" onClick={this.handleStartAgainClick} />
          <div className={cx(style.action, style.feedbackBar)}>
            <I18n component="div" stringId="results.feedbackPreface" />
            <I18n
              component="a"
              stringId="results.feedbackLinkText"
              href={`${feedbackPrefillUrl}${resultData.testId}`}
              target="_blank"
              rel="noopener"
            />
          </div>
        </div>
      </div>
    );
  }
}

export { Results as Pure };

export default connect(mapSelectors({
  scoreStatistics,
  resultData,
  missedCards
}), { resetTest })(Results);
