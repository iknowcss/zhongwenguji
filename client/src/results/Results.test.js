import React from 'react';
import Renderer from 'react-test-renderer';
import { hasClassName } from '../util/asdf.testutil';
import ResultsDisplay from './ResultsDisplay';
import ResultsChart from './ResultsChart';
import { Pure as Results } from './Results';

jest.mock('./ResultsDisplay', () => () => <div className="ResultsDisplay" />);
jest.mock('./ResultsChart', () => () => <div className="ResultsChart" />);

describe('Results', () => {
  it('renders', () => {
    const component = Renderer.create(<Results />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('passes results data down to children', () => {
    const knownEstimate = 1500;
    const knownEstimateUncertainty = 200;
    const curvePoints = [[0, 53], [249, 53], [498, 53]];
    const samplePoints = [[124.5, 60], [373.5, 40], [622.5, 60]];
    const resultData = { knownEstimate, knownEstimateUncertainty, curvePoints, samplePoints };
    const component = Renderer.create(<Results resultData={resultData} />);

    const resultsDisplay = component.root.findByType(ResultsDisplay);
    expect(resultsDisplay.props).toEqual(
      expect.objectContaining({
        estimate: knownEstimate,
        uncertainty: knownEstimateUncertainty
      })
    );

    const resultsChart = component.root.findByType(ResultsChart);
    expect(resultsChart.props).toEqual(
      expect.objectContaining({ curvePoints, samplePoints })
    );
  });

  it('starts again', () => {
    const preventDefault = jest.fn();
    const resetTest = jest.fn();
    const component = Renderer.create(<Results resetTest={resetTest} />);
    const startAgainButton = component.root.findByProps({ stringId: 'results.startAgain' });
    startAgainButton.props.onClick({ preventDefault });
    expect(preventDefault).toHaveBeenCalledTimes(1);
    expect(resetTest).toHaveBeenCalledTimes(1);
  });

  it('reviews missed characters', () => {
    const preventDefault = jest.fn();
    const reviewMissed = jest.fn();
    const component = Renderer.create(<Results reviewMissed={reviewMissed} />);
    const reviewMissedButton = component.root.findByProps({ stringId: 'results.reviewMissedCharacters' });
    reviewMissedButton.props.onClick({ preventDefault });
    expect(preventDefault).toHaveBeenCalledTimes(1);
    expect(reviewMissed).toHaveBeenCalledTimes(1);
  });
});
