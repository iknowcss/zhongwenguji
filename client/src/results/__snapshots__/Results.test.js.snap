// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Results renders 1`] = `
<div
  className="container"
>
  <div
    className="ResultsDisplay"
  />
  <div
    className="ResultsChart"
  />
  <div
    className="actionContainer"
  >
    <button
      className="button action"
    >
      results.reviewMissedCharacters
    </button>
    <button
      className="button accent action"
      onClick={[Function]}
    >
      results.startAgain
    </button>
    <div
      className="action feedbackBar"
    >
      <div>
        results.feedbackPreface
      </div>
      <a
        href="https://docs.google.com/forms/d/e/1FAIpQLSe74g0vXWUuVsQ9uqYGQhLFADOeqiUHfyyjBqT4AfdY4VOYxg/viewform?usp=pp_url&entry.1889314490=undefined"
        rel="noopener"
        target="_blank"
      >
        results.feedbackLinkText
      </a>
    </div>
  </div>
</div>
`;
